### DISCUSSION TITLE
Entrypoint scripts help

### DISCUSSION THREAD

**Discussion by dpletiko** [serversideup/docker-php#359] [Q&A] [answered]
Hello everyone, I've been trying to play around with this repo (Cool one, btw.), to see how it works with Laravel and so far I kinda got stuck on understanding entrypoint.d scripts. Still in process of learning docker...

Let's assume I have a few small .sh scripts that should run on container start and that the default one "laravel-automations" is/will be disabled.

Can anyone help me realize how to handle this using s6-init and without s6, using only entrypoint?

I would like to know how, if possible, to wait until all of the entrypoint.d scripts are finished to start the container / nginx / fpm? Unless I'm mistaken / went down the wrong path.

Because during the execution of the custom scripts I get messages like this randomly spread out:
```
[13-May-2024 12:36:00] NOTICE: fpm is running, pid 126                           
[13-May-2024 12:36:00] NOTICE: ready to handle connections  
...
127.0.0.1 -  13/May/2024:12:36:01 +0000 "GET /status" 200                        
127.0.0.1 -  13/May/2024:12:36:01 +0000 "GET /ping" 200    
```

### Dockerfile:
```
FROM serversideup/php:beta-8.3-fpm-nginx-bookworm as base

ENV S6_CMD_WAIT_FOR_SERVICES=1
ENV S6_CMD_WAIT_FOR_SERVICES_MAXTIME=0
COPY --chmod=755 docker/entrypoint.d/ /etc/entrypoint.d/

ENV AUTORUN_ENABLED=false

ARG USER=www-data

# Switch to root so we can do root things
USER root

RUN docker-php-serversideup-dep-install-debian \
    apt-utils,gettext,git,netcat-traditional,procps,vim;

# Install the intl extension with root permissions
RUN install-php-extensions \
    calendar \
    exif \
    gd \
    intl \
    opcache \
    pdo_pgsql \
    zip;


RUN docker-php-serversideup-s6-init

#ENTRYPOINT docker-php-serversideup-entrypoint

USER ${USER}

FROM base as deploy

COPY --chown=${USER}:${USER} . "$APP_BASE_DIR"

USER ${USER}

WORKDIR $APP_BASE_DIR

RUN composer install \
    --prefer-dist \
    --no-progress \
    --optimize-autoloader \
    --no-dev \
    --no-interaction \
    --working-dir="$APP_BASE_DIR";
```

Thanks for help!

**Comment by jaydrogers** [comment]
I think I understand your question, but please rephrase your question if I didn't answer correctly.

> Let's assume I have a few small .sh scripts that should run on container start and that the default one "laravel-automations" is/will be disabled. Can anyone help me realize how to handle this using s6-init and without s6, using only entrypoint?

Make sure you have all scripts in order. You can see our scripts by running `ls -al /etc/entrypoint.d`:

```
drwxr-xr-x 1 root root 4096 May 13 13:05 .
drwxr-xr-x 1 root root 4096 May 13 13:43 ..
-rwxr-xr-x 1 root root 1600 May 13 13:05 0-container-info.sh
-rwxr-xr-x 1 root root 1216 May 13 13:05 1-debug-mode.sh
-rwxr-xr-x 1 root root 6307 May 13 13:05 10-init-webserver-config.sh
-rwxr-xr-x 1 root root 5225 May 13 13:05 50-laravel-automations.sh
```

So if you want your scripts to run after ours, just start them with any number greater than 50.

More information here: https://serversideup.net/open-source/docker-php/docs/customizing-the-image/adding-your-own-start-up-scripts#choose-your-execution-order

**Reply by dpletiko** [reply]
Yes, I already named all the scripts in order (assuming the order starts after 50-laravel-automations), but as far as I can understand, the issue is that the "nginx" depends on "php-fpm" but the "php-fpm" depends on "50-laravel-automations".

So the nginx and php-fpm start before any of the entrypoint scripts named/ordered after "50-" prefix...

I hope I make some sense.

Thanks

**Reply by jaydrogers** [reply, marked as answer]
Oh, I see now.

With those advanced dependencies, you might want to write your own service script: https://github.com/just-containers/s6-overlay/tree/master#writing-a-service-script

You wont need to use my `docker-php-serversideup-s6-init`. You can see how I structured my services here: https://github.com/serversideup/docker-php/tree/main/src/variations/fpm-nginx/etc/s6-overlay/s6-rc.d/nginx

**Reply by dpletiko** [reply]
Oh, I understand now. I also realized that the base entrypoints/s6-rc are already bundled inside the image.

The deps I have are not really advanced, but actually split into few files, with output messages, like envsubst, mkdirs, laravel db/artisan stuff,...

Would be awesome if the php-fpm depencency could be configured during build so the s6-init could still be used or even disabled so the entrypoint sh scripts could take the job if needed. Then the .sh scripts could be "generic"...

Or maybe have the s6-init script handle the setup so it can include custom entrypoint.d files?

**Reply by jaydrogers** [reply]
We can definitely explore that area. We'd just need to put together a thought out feature request on how to solve it: https://github.com/serversideup/docker-php/discussions/66

My biggest thing would be finding the line of when does a user just write their own script vs having the script build it for them.

For something like `docker-php-serversideup-s6-init`, it was mainly a tool written for the maintainers of the images so we could easily solve the problem of writing one entrypoint script and having it convert to S6.

For external use cases, it might just be easier for the user to add their own S6 service like we did.

**Reply by dpletiko** [reply]
Awesome, let's see what happens...

The `s6-init` script is helpful, at least was for me to somewhat faster understand what is happening under the hood, so I can see your point there.

The only thing, if I haven't missed something of course, is that to tweak the s6 setup (at least this slightly) we can fully `DISABLE_DEFAULT_CONFIG` which feels like an overkill or we can edit the already present files in the container which assumes we won't stumble upon a "breaking change"...

**Reply by jaydrogers** [reply]
Correct! If you want to remove any of our functionality, I would advise to DISABLE_DEFAULT_CONFIG

If you want to keep our functionality, but then just add your own script after -- just add it as a dependency.

Regardless, I would advise to subscribe to repo updates so you're notified on releases and changes: https://serversideup.net/open-source/docker-php/docs/getting-started/upgrade-guide#subscribe-to-repository-updates

**Reply by dpletiko** [reply]
Thanks for the info, I started to write custom services to get the hang of it.

I'd just like to know, since the `USER` is set in `Dockerfile` to `www-data` and is used for `fpm`/`nginx` services, is there a way to write a service file that uses the `root` user for needed permissions?

Considering `s6-overlay` only has a way to drop privileges... Unless I missed something.

Thanks :smile:

**Reply by jaydrogers** [reply]
Yeah you can definitely switch permissions. See our example file on how we jump back and forth: https://serversideup.net/open-source/docker-php/docs/guide/migrating-from-v2-to-v3#cicd

S6 does have [their own way of dropping privileges](https://github.com/just-containers/s6-overlay?tab=readme-ov-file#dropping-privileges), but they also have been moving to the [Docker Standard with USER](https://github.com/just-containers/s6-overlay?tab=readme-ov-file#user-directive)
